header-rewrite-filter/header_processor.cc

```diff
@@ -250,14 +259,72 @@ namespace HeaderRewriteFilter {
             return absl::InvalidArgumentError("error parsing boolean expression -- " + std::string(e.what()));
         }
 
+
         return absl::OkStatus();
     }
 
-    // return status and bool result
-    std::tuple<absl::Status, bool> SetBoolProcessor::executeOperation(bool negate) {
+    std::tuple<absl::Status, bool> SetBoolProcessor::executeOperation(Http::RequestOrResponseHeaderMap& headers, bool negate) {
+        bool result;
+        std::string source; // result of calling a function to fetch req/res metadata
+        switch (function_type_) {
+            case Utility::FunctionType::GetHdr:
+                try {
+                    const auto arguments = StringUtil::splitToken(function_argument_, ",", false, true);
+
+                    Http::LowerCaseString header_key(arguments.at(0));
+                    const auto num_header_entries = headers.get((header_key)).size();
+                    if (num_header_entries == 0) { // header does not exist
+                        result = false;
+                        bool apply_negation = negate ? !result : result;
+                        return std::make_tuple(absl::OkStatus(), apply_negation);
+                    }
+
+                    // get header value based on an optional position argument, if no position specified return the last value
+                    if (arguments.size() > 2) {
+                        return std::make_tuple(absl::InvalidArgumentError("invalid match syntax -- too many arguments to hdr function"), false);
+                    }
+
+                    int index = std::stoi(std::string(arguments.at(1)));
+                    if (arguments.size() == 1 || index == -1) { // no position specified or last argument specified
+                        index = num_header_entries - 1;
+                    }
+
+                    // validate index
+                    if (index < 0 || index >= num_header_entries) {
+                        return std::make_tuple(absl::InvalidArgumentError("invalid match syntax -- hdr position out of bounds"), false);
+                    }
+
+                    // get last comma-separated value of the header
+                    const auto header_entry = headers.get((header_key))[index]; // must be indexed with brackets
+                    const auto header_val = header_entry->value().getStringView();
+                    source = std::string(header_val);
+                } catch (std::exception& e) {
+                    return std::make_tuple(absl::InvalidArgumentError("failed to perform boolean match" + std::string(e.what())), false);
+                }
+                break;
+            case Utility::FunctionType::Urlp:
+            {
+                try {
+                    Http::RequestHeaderMap* request_headers = dynamic_cast<Http::RequestHeaderMap*>(&headers); // can fail if invalid config is provided, ie if response tries to get path
```

Comment:
Might need to know whether it's a request or a response since this currently doesn't fail gracefully.

Reply:
Yeah, I think this should ideally figure out from the parser whether it's a request or response. You should be able to pass that in somehow from the Envoy event. 

If that isn't possible, `dynamic_cast` will return a null pointer, and you can check for null here.
Otherwise the `request_headers->Path()` below will crash (Undefined Behaviour).

Is a `urlp` valid on a Response hook? Maybe the config should be rejected when parsing if it's `http-response ... urlp`?

Reply:
Even if we do reject that in parsing, doing a null check and returning and error here would be good, just in case there's a bug somewhere and this somehow gets called on a response hook.

C++ exceptions are very slow, relative to the speed of HTTP requests in Production. Doing the null check will ensure Envoy keeps running, whereas if a bug causes an exception to be thrown on every request, it could potentially bring down the system